File: packages/swift/Sources/LumenUI/Primitives.swift
import SwiftUI

public enum LumenTextTone: Sendable {
    case danger
    case `default`
    case muted
    case soft
    case success
    case warning
}

public enum LumenTextVariant: Sendable {
    case body
    case caption
    case label
    case title
}

public struct LumenText: View {
    @Environment(\.lumenTheme) private var theme

    private let content: LumenTextContent
    private let tone: LumenTextTone
    private let variant: LumenTextVariant

    public init(
        _ content: LocalizedStringKey,
        variant: LumenTextVariant = .body,
        tone: LumenTextTone = .default
    ) {
        self.content = .localized(content)
        self.variant = variant
        self.tone = tone
    }

    public init(
        _ content: LumenTextContent,
        variant: LumenTextVariant = .body,
        tone: LumenTextTone = .default
    ) {
        self.content = content
        self.variant = variant
        self.tone = tone
    }

    public var body: some View {
        content.text
            .font(font)
            .foregroundStyle(foregroundColor)
    }

    private var font: Font {
        switch variant {
        case .body: .body
        case .caption: .caption
        case .label: .callout.weight(.semibold)
        case .title: .title2.weight(.bold)
        }
    }

    private var foregroundColor: Color {
        switch tone {
        case .danger: theme.colors.danger
        case .default: theme.colors.ink
        case .muted: theme.colors.inkMuted
        case .soft: theme.colors.inkSoft
        case .success: theme.colors.success
        case .warning: theme.colors.warning
        }
    }
}

public enum LumenSurfacePadding: Sendable {
    case lg
    case md
    case none
    case sm
    case xl

    var value: CGFloat {
        switch self {
        case .lg: LumenSpacing.lg
        case .md: LumenSpacing.md
        case .none: 0
        case .sm: LumenSpacing.sm
        case .xl: LumenSpacing.xl
        }
    }
}

public enum LumenSurfaceRadius: Sendable {
    case size2xl
    case size3xl
    case lg
    case md
    case none
    case sm
    case xl

    var value: CGFloat {
        switch self {
        case .size2xl: LumenRadius.size2xl
        case .size3xl: LumenRadius.size3xl
        case .lg: LumenRadius.lg
        case .md: LumenRadius.md
        case .none: 0
        case .sm: LumenRadius.sm
        case .xl: LumenRadius.xl
        }
    }
}

public enum LumenSurfaceTone: Sendable {
    case canvas
    case muted
    case strong
    case surface
}

public struct LumenSurface<Content: View>: View {
    @Environment(\.lumenTheme) private var theme

    private let content: Content
    private let padding: LumenSurfacePadding
    private let radius: LumenSurfaceRadius
    private let tone: LumenSurfaceTone

    public init(
        tone: LumenSurfaceTone = .surface,
        padding: LumenSurfacePadding = .md,
        radius: LumenSurfaceRadius = .md,
        @ViewBuilder content: () -> Content
    ) {
        self.content = content()
        self.padding = padding
        self.radius = radius
        self.tone = tone
    }

    public var body: some View {
        content
            .padding(padding.value)
            .background(backgroundColor)
            .clipShape(RoundedRectangle(cornerRadius: radius.value, style: .continuous))
    }

    private var backgroundColor: Color {
        switch tone {
        case .canvas: theme.colors.canvas
        case .muted: theme.colors.surfaceMuted
        case .strong: theme.colors.surfaceStrong
        case .surface: theme.colors.surface
        }
    }
}

public enum LumenButtonIntent: Sendable {
    case danger
    case primary
    case quiet
    case secondary
}

public enum LumenControlSize: Sendable {
    case lg
    case md
    case sm
}

public enum LumenIconSize: Sendable {
    case lg
    case md
    case sm

    var dimension: CGFloat {
        switch self {
        case .lg: 24
        case .md: 20
        case .sm: 16
        }
    }
}

public struct LumenIcon: View {
    @Environment(\.lumenTheme) private var theme

    private let color: Color?
    private let label: LocalizedStringKey?
    private let name: LumenIconName?
    private let size: LumenIconSize
    private let systemName: String?

    public init(
        name: LumenIconName,
        size: LumenIconSize = .md,
        color: Color? = nil,
        label: LocalizedStringKey? = nil
    ) {
        self.name = name
        self.systemName = nil
        self.size = size
        self.color = color
        self.label = label
    }

    public init(
        systemName: String,
        size: LumenIconSize = .md,
        color: Color? = nil,
        label: LocalizedStringKey? = nil
    ) {
        self.name = nil
        self.systemName = systemName
        self.size = size
        self.color = color
        self.label = label
    }

    public var body: some View {
        iconImage
            .foregroundStyle(color ?? theme.colors.ink)
            .frame(width: size.dimension, height: size.dimension)
            .accessibilityHidden(label == nil)
            .accessibilityLabel(label ?? "")
    }

    @ViewBuilder private var iconImage: some View {
        if let name {
            Image(name.assetName, bundle: .module)
                .resizable()
                .scaledToFit()
        } else if let systemName {
            Image(systemName: systemName)
                .font(.system(size: size.dimension, weight: .medium))
        }
    }
}

public struct LumenIconButtonMetrics: Equatable, Sendable {
    public let iconSize: LumenIconSize
    public let touchTarget: CGFloat

    public static func resolve(
        _ size: LumenControlSize,
        density: LumenControlDensity = .platformDefault
    ) -> LumenIconButtonMetrics {
        switch (density, size) {
        case (.compact, .lg): LumenIconButtonMetrics(iconSize: .lg, touchTarget: 40)
        case (.compact, .md): LumenIconButtonMetrics(iconSize: .md, touchTarget: 32)
        case (.compact, .sm): LumenIconButtonMetrics(iconSize: .sm, touchTarget: 28)
        case (.regular, .lg): LumenIconButtonMetrics(iconSize: .lg, touchTarget: 52)
        case (.regular, .md): LumenIconButtonMetrics(iconSize: .md, touchTarget: 44)
        case (.regular, .sm): LumenIconButtonMetrics(iconSize: .sm, touchTarget: 44)
        }
    }
}

public struct LumenIconButtonStyle: ButtonStyle {
    @Environment(\.lumenControlDensity) private var density
    @Environment(\.isEnabled) private var isEnabled
    @Environment(\.lumenTheme) private var theme

    private let intent: LumenButtonIntent
    private let size: LumenControlSize

    public init(intent: LumenButtonIntent = .quiet, size: LumenControlSize = .md) {
        self.intent = intent
        self.size = size
    }

    public func makeBody(configuration: Configuration) -> some View {
        let metrics = LumenIconButtonMetrics.resolve(size, density: density)

        configuration.label
            .frame(width: metrics.touchTarget, height: metrics.touchTarget)
            .background(backgroundColor)
            .overlay {
                RoundedRectangle(cornerRadius: LumenRadius.sm, style: .continuous)
                    .stroke(borderColor, lineWidth: 1)
            }
            .clipShape(RoundedRectangle(cornerRadius: LumenRadius.sm, style: .continuous))
            .opacity(isEnabled ? (configuration.isPressed ? 0.84 : 1) : 0.52)
    }

    private var backgroundColor: Color {
        switch intent {
        case .danger: theme.colors.danger
        case .primary: theme.colors.brandSolid
        case .quiet: .clear
        case .secondary: theme.colors.surfaceMuted
        }
    }

    private var borderColor: Color {
        switch intent {
        case .danger: theme.colors.danger
        case .primary: theme.colors.brandSolid
        case .quiet: .clear
        case .secondary: theme.colors.line
        }
    }
}

public struct LumenIconButton: View {
    @Environment(\.lumenControlDensity) private var density
    @Environment(\.lumenTheme) private var theme

    private let action: () -> Void
    private let disabled: Bool
    private let intent: LumenButtonIntent
    private let label: LocalizedStringKey
    private let name: LumenIconName?
    private let size: LumenControlSize
    private let systemName: String?

    public init(
        name: LumenIconName,
        label: LocalizedStringKey,
        intent: LumenButtonIntent = .quiet,
        size: LumenControlSize = .md,
        disabled: Bool = false,
        action: @escaping () -> Void
    ) {
        self.name = name
        self.systemName = nil
        self.label = label
        self.intent = intent
        self.size = size
        self.disabled = disabled
        self.action = action
    }

    public init(
        systemName: String,
        label: LocalizedStringKey,
        intent: LumenButtonIntent = .quiet,
        size: LumenControlSize = .md,
        disabled: Bool = false,
        action: @escaping () -> Void
    ) {
        self.name = nil
        self.systemName = systemName
        self.label = label
        self.intent = intent
        self.size = size
        self.disabled = disabled
        self.action = action
    }

    public var body: some View {
        let metrics = LumenIconButtonMetrics.resolve(size, density: density)

        Button(action: action) {
            icon(size: metrics.iconSize)
        }
        .buttonStyle(LumenIconButtonStyle(intent: intent, size: size))
        .disabled(disabled)
        .accessibilityLabel(label)
    }

    @ViewBuilder private func icon(size: LumenIconSize) -> some View {
        if let name {
            LumenIcon(name: name, size: size, color: foregroundColor)
        } else if let systemName {
            LumenIcon(systemName: systemName, size: size, color: foregroundColor)
        }
    }

    private var foregroundColor: Color {
        switch intent {
        case .danger: theme.colors.onDanger
        case .primary: theme.colors.onBrand
        case .quiet: theme.colors.inkSoft
        case .secondary: theme.colors.ink
        }
    }
}

public struct LumenButtonMetrics: Equatable, Sendable {
    public let horizontalPadding: CGFloat
    public let minHeight: CGFloat

    public static func resolve(
        _ size: LumenControlSize,
        density: LumenControlDensity = .platformDefault
    ) -> LumenButtonMetrics {
        switch (density, size) {
        case (.compact, .lg):
            LumenButtonMetrics(horizontalPadding: LumenSpacing.lg, minHeight: 40)
        case (.compact, .md):
            LumenButtonMetrics(horizontalPadding: LumenSpacing.md, minHeight: 32)
        case (.compact, .sm):
            LumenButtonMetrics(horizontalPadding: LumenSpacing.md, minHeight: 28)
        case (.regular, .lg):
            LumenButtonMetrics(horizontalPadding: LumenSpacing.xl, minHeight: 52)
        case (.regular, .md):
            LumenButtonMetrics(horizontalPadding: LumenSpacing.lg, minHeight: 44)
        case (.regular, .sm):
            LumenButtonMetrics(horizontalPadding: LumenSpacing.md, minHeight: 36)
        }
    }
}

public struct LumenButtonStyle: ButtonStyle {
    @Environment(\.lumenControlDensity) private var density
    @Environment(\.isEnabled) private var isEnabled
    @Environment(\.lumenTheme) private var theme

    private let intent: LumenButtonIntent
    private let size: LumenControlSize

    public init(
        intent: LumenButtonIntent = .primary,
        size: LumenControlSize = .md
    ) {
        self.intent = intent
        self.size = size
    }

    public func makeBody(configuration: Configuration) -> some View {
        let metrics = LumenButtonMetrics.resolve(size, density: density)

        configuration.label
            .font(size == .lg ? .body.weight(.semibold) : .callout.weight(.semibold))
            .foregroundStyle(foregroundColor)
            .frame(minHeight: metrics.minHeight)
            .padding(.horizontal, metrics.horizontalPadding)
            .background(backgroundColor)
            .overlay {
                RoundedRectangle(cornerRadius: LumenRadius.sm, style: .continuous)
                    .stroke(borderColor, lineWidth: 1)
            }
            .clipShape(RoundedRectangle(cornerRadius: LumenRadius.sm, style: .continuous))
            .opacity(isEnabled ? (configuration.isPressed ? 0.84 : 1) : 0.52)
    }

    private var backgroundColor: Color {
        switch intent {
        case .danger: theme.colors.danger
        case .primary: theme.colors.brandSolid
        case .quiet: .clear
        case .secondary: theme.colors.surfaceMuted
        }
    }

    private var borderColor: Color {
        switch intent {
        case .danger: theme.colors.danger
        case .primary: theme.colors.brandSolid
        case .quiet: .clear
        case .secondary: theme.colors.line
        }
    }

    private var foregroundColor: Color {
        switch intent {
        case .danger: theme.colors.onDanger
        case .primary: theme.colors.onBrand
        case .quiet: theme.colors.inkSoft
        case .secondary: theme.colors.ink
        }
    }
}

public struct LumenButton<Label: View>: View {
    private let action: () -> Void
    private let disabled: Bool
    private let intent: LumenButtonIntent
    private let label: Label
    private let loading: Bool
    private let size: LumenControlSize

    public init(
        intent: LumenButtonIntent = .primary,
        size: LumenControlSize = .md,
        loading: Bool = false,
        disabled: Bool = false,
        action: @escaping () -> Void,
        @ViewBuilder label: () -> Label
    ) {
        self.action = action
        self.disabled = disabled
        self.intent = intent
        self.label = label()
        self.loading = loading
        self.size = size
    }

    public var body: some View {
        Button(action: action) {
            HStack(spacing: LumenSpacing.sm) {
                if loading {
#if os(tvOS)
                    ProgressView()
#else
                    ProgressView()
                        .controlSize(.small)
#endif
                }
                label
            }
        }
        .buttonStyle(LumenButtonStyle(intent: intent, size: size))
        .disabled(disabled || loading)
        .accessibilityValue(loading ? "Loading" : "")
    }
}

public extension LumenButton where Label == Text {
    init(
        _ title: LocalizedStringKey,
        intent: LumenButtonIntent = .primary,
        size: LumenControlSize = .md,
        loading: Bool = false,
        disabled: Bool = false,
        action: @escaping () -> Void
    ) {
        self.init(
            intent: intent,
            size: size,
            loading: loading,
            disabled: disabled,
            action: action
        ) {
            Text(title)
        }
    }

    init(
        _ title: LumenTextContent,
        intent: LumenButtonIntent = .primary,
        size: LumenControlSize = .md,
        loading: Bool = false,
        disabled: Bool = false,
        action: @escaping () -> Void
    ) {
        self.init(
            intent: intent,
            size: size,
            loading: loading,
            disabled: disabled,
            action: action
        ) {
            title.text
        }
    }
}

public struct LumenTextField: View {
    @Binding private var text: String
    @Environment(\.isEnabled) private var isEnabled
    @Environment(\.lumenControlDensity) private var density
    @Environment(\.lumenTheme) private var theme

    private let error: Bool
    private let errorMessage: LumenTextContent?
    private let size: LumenControlSize
    private let title: LumenTextContent

    public init(
        _ title: String,
        text: Binding<String>,
        size: LumenControlSize = .md,
        error: Bool = false,
        errorMessage: String? = nil
    ) {
        self.title = .verbatim(title)
        _text = text
        self.size = size
        self.error = error
        self.errorMessage = errorMessage.map(LumenTextContent.verbatim)
    }

    public init(
        _ title: LumenTextContent,
        text: Binding<String>,
        size: LumenControlSize = .md,
        error: Bool = false,
        errorMessage: LumenTextContent? = nil
    ) {
        self.title = title
        _text = text
        self.size = size
        self.error = error
        self.errorMessage = errorMessage
    }

    public var body: some View {
        let validationMessage = error || errorMessage != nil
            ? errorMessage ?? .localized(LocalizedStringKey("Invalid value"))
            : nil

        TextField(text: $text) {
            title.text
        }
            .textFieldStyle(.plain)
            .font(size == .lg ? .body : .callout)
            .foregroundStyle(theme.colors.ink)
            .frame(minHeight: LumenButtonMetrics.resolve(size, density: density).minHeight)
            .padding(.horizontal, LumenSpacing.md)
            .background(theme.colors.surface)
            .overlay {
                RoundedRectangle(cornerRadius: LumenRadius.sm, style: .continuous)
                    .stroke(validationMessage == nil ? theme.colors.line : theme.colors.danger, lineWidth: 1)
            }
            .clipShape(RoundedRectangle(cornerRadius: LumenRadius.sm, style: .continuous))
            .opacity(isEnabled ? 1 : 0.52)
            .lumenAccessibilityHint(validationMessage)
            .accessibilityValue(Text(verbatim: text))
    }
}

public enum LumenBadgeTone: Sendable {
    case accent
    case danger
    case neutral
    case success
    case warning
}

public struct LumenBadge: View {
    @Environment(\.lumenTheme) private var theme

    private let content: LumenTextContent
    private let tone: LumenBadgeTone

    public init(
        _ content: LocalizedStringKey,
        tone: LumenBadgeTone = .neutral
    ) {
        self.content = .localized(content)
        self.tone = tone
    }

    public init(
        _ content: LumenTextContent,
        tone: LumenBadgeTone = .neutral
    ) {
        self.content = content
        self.tone = tone
    }

    public var body: some View {
        content.text
            .font(.caption.weight(.semibold))
            .foregroundStyle(foregroundColor)
            .padding(.horizontal, LumenSpacing.sm)
            .padding(.vertical, LumenSpacing.xs)
            .background(backgroundColor)
            .clipShape(Capsule())
    }

    private var foregroundColor: Color {
        switch tone {
        case .accent: theme.colors.accent
        case .danger: theme.colors.danger
        case .neutral: theme.colors.inkSoft
        case .success: theme.colors.success
        case .warning: theme.colors.warning
        }
    }

    private var backgroundColor: Color {
        tone == .neutral ? theme.colors.surfaceMuted : foregroundColor.opacity(0.12)
    }
}

public struct LumenDivider: View {
    @Environment(\.lumenTheme) private var theme

    public init() {}

    public var body: some View {
        Rectangle()
            .fill(theme.colors.line)
            .frame(height: 1)
            .accessibilityHidden(true)
    }
}

public struct LumenSpinner: View {
    @Environment(\.lumenTheme) private var theme

    private let label: LumenTextContent

    public init(_ label: LocalizedStringKey = "Loading") {
        self.label = .localized(label)
    }

    public init(_ label: LumenTextContent) {
        self.label = label
    }

    public var body: some View {
        ProgressView()
            .tint(theme.colors.brand)
            .accessibilityLabel(label.text)
    }
}

File: packages/compose/src/main/kotlin/com/santi020k/lumen/Primitives.kt
package com.santi020k.lumen

import androidx.compose.foundation.BorderStroke
import androidx.compose.foundation.border
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.PaddingValues
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.RowScope
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.defaultMinSize
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.Button
import androidx.compose.material3.ButtonDefaults
import androidx.compose.material3.CircularProgressIndicator
import androidx.compose.material3.HorizontalDivider
import androidx.compose.material3.Icon
import androidx.compose.material3.IconButton
import androidx.compose.material3.IconButtonDefaults
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.OutlinedTextField
import androidx.compose.material3.OutlinedTextFieldDefaults
import androidx.compose.material3.Surface
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.Immutable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.vector.ImageVector
import androidx.compose.ui.draw.clip
import androidx.compose.ui.res.painterResource
import androidx.compose.ui.semantics.contentDescription
import androidx.compose.ui.semantics.semantics
import androidx.compose.ui.semantics.stateDescription
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.input.VisualTransformation
import androidx.compose.ui.unit.Dp
import androidx.compose.ui.unit.dp

enum class LumenTextTone {
    Danger,
    Default,
    Muted,
    Soft,
    Success,
    Warning
}

enum class LumenTextVariant {
    Body,
    Caption,
    Label,
    Title
}

@Composable
fun LumenText(
    text: String,
    modifier: Modifier = Modifier,
    variant: LumenTextVariant = LumenTextVariant.Body,
    tone: LumenTextTone = LumenTextTone.Default
) {
    val theme = LocalLumenTheme.current
    val color = when (tone) {
        LumenTextTone.Danger -> theme.colors.danger
        LumenTextTone.Default -> theme.colors.ink
        LumenTextTone.Muted -> theme.colors.inkMuted
        LumenTextTone.Soft -> theme.colors.inkSoft
        LumenTextTone.Success -> theme.colors.success
        LumenTextTone.Warning -> theme.colors.warning
    }
    val style = when (variant) {
        LumenTextVariant.Body -> MaterialTheme.typography.bodyLarge
        LumenTextVariant.Caption -> MaterialTheme.typography.bodySmall
        LumenTextVariant.Label -> MaterialTheme.typography.labelLarge.copy(fontWeight = FontWeight.SemiBold)
        LumenTextVariant.Title -> MaterialTheme.typography.headlineSmall.copy(fontWeight = FontWeight.Bold)
    }

    Text(text = text, modifier = modifier, color = color, style = style)
}

enum class LumenSurfacePadding(val value: Dp) {
    None(0.dp),
    Sm(LumenSpacing.Sm),
    Md(LumenSpacing.Md),
    Lg(LumenSpacing.Lg),
    Xl(LumenSpacing.Xl)
}

enum class LumenSurfaceRadius(val value: Dp) {
    None(0.dp),
    Sm(LumenRadius.Sm),
    Md(LumenRadius.Md),
    Lg(LumenRadius.Lg),
    Xl(LumenRadius.Xl),
    Size2xl(LumenRadius.Size2xl),
    Size3xl(LumenRadius.Size3xl)
}

enum class LumenSurfaceTone {
    Canvas,
    Muted,
    Strong,
    Surface
}

@Composable
fun LumenSurface(
    modifier: Modifier = Modifier,
    tone: LumenSurfaceTone = LumenSurfaceTone.Surface,
    padding: LumenSurfacePadding = LumenSurfacePadding.Md,
    radius: LumenSurfaceRadius = LumenSurfaceRadius.Md,
    content: @Composable () -> Unit
) {
    val colors = LocalLumenTheme.current.colors
    val color = when (tone) {
        LumenSurfaceTone.Canvas -> colors.canvas
        LumenSurfaceTone.Muted -> colors.surfaceMuted
        LumenSurfaceTone.Strong -> colors.surfaceStrong
        LumenSurfaceTone.Surface -> colors.surface
    }

    Surface(
        modifier = modifier,
        color = color,
        shape = RoundedCornerShape(radius.value)
    ) {
        androidx.compose.foundation.layout.Box(
            modifier = Modifier.padding(padding.value),
            contentAlignment = Alignment.TopStart
        ) {
            content()
        }
    }
}

enum class LumenButtonIntent {
    Danger,
    Primary,
    Quiet,
    Secondary
}

enum class LumenControlSize {
    Lg,
    Md,
    Sm
}

enum class LumenIconSize(val dimension: Dp) {
    Sm(16.dp),
    Md(20.dp),
    Lg(24.dp)
}

@Composable
fun LumenIcon(
    imageVector: ImageVector,
    modifier: Modifier = Modifier,
    contentDescription: String? = null,
    size: LumenIconSize = LumenIconSize.Md,
    tint: Color = LocalLumenTheme.current.colors.ink
) {
    Icon(
        imageVector = imageVector,
        contentDescription = contentDescription,
        modifier = modifier.size(size.dimension),
        tint = tint
    )
}

@Composable
fun LumenIcon(
    name: LumenIconName,
    modifier: Modifier = Modifier,
    contentDescription: String? = null,
    size: LumenIconSize = LumenIconSize.Md,
    tint: Color = LocalLumenTheme.current.colors.ink
) {
    Icon(
        painter = painterResource(name.resourceId),
        contentDescription = contentDescription,
        modifier = modifier.size(size.dimension),
        tint = tint
    )
}

@Immutable
data class LumenIconButtonMetrics(
    val iconSize: LumenIconSize,
    val touchTarget: Dp
) {
    companion object {
        fun resolve(size: LumenControlSize): LumenIconButtonMetrics = when (size) {
            LumenControlSize.Lg -> LumenIconButtonMetrics(LumenIconSize.Lg, 52.dp)
            LumenControlSize.Md -> LumenIconButtonMetrics(LumenIconSize.Md, 44.dp)
            LumenControlSize.Sm -> LumenIconButtonMetrics(LumenIconSize.Sm, 44.dp)
        }
    }
}

@Composable
fun LumenIconButton(
    imageVector: ImageVector,
    contentDescription: String,
    onClick: () -> Unit,
    modifier: Modifier = Modifier,
    intent: LumenButtonIntent = LumenButtonIntent.Quiet,
    size: LumenControlSize = LumenControlSize.Md,
    enabled: Boolean = true
) {
    val palette = buttonPalette(LocalLumenTheme.current.colors, intent)
    val metrics = LumenIconButtonMetrics.resolve(size)
    val shape = RoundedCornerShape(LumenRadius.Sm)

    IconButton(
        onClick = onClick,
        modifier = modifier
            .size(metrics.touchTarget)
            .clip(shape)
            .border(1.dp, palette.border, shape),
        enabled = enabled,
        colors = IconButtonDefaults.iconButtonColors(
            containerColor = palette.background,
            contentColor = palette.foreground,
            disabledContainerColor = palette.background.copy(alpha = 0.52f),
            disabledContentColor = palette.foreground.copy(alpha = 0.52f)
        )
    ) {
        LumenIcon(
            imageVector = imageVector,
            contentDescription = contentDescription,
            size = metrics.iconSize,
            tint = palette.foreground.copy(alpha = if (enabled) 1f else 0.52f)
        )
    }
}

@Composable
fun LumenIconButton(
    name: LumenIconName,
    contentDescription: String,
    onClick: () -> Unit,
    modifier: Modifier = Modifier,
    intent: LumenButtonIntent = LumenButtonIntent.Quiet,
    size: LumenControlSize = LumenControlSize.Md,
    enabled: Boolean = true
) {
    val palette = buttonPalette(LocalLumenTheme.current.colors, intent)
    val metrics = LumenIconButtonMetrics.resolve(size)
    val shape = RoundedCornerShape(LumenRadius.Sm)

    IconButton(
        onClick = onClick,
        modifier = modifier
            .size(metrics.touchTarget)
            .clip(shape)
            .border(1.dp, palette.border, shape),
        enabled = enabled,
        colors = IconButtonDefaults.iconButtonColors(
            containerColor = palette.background,
            contentColor = palette.foreground,
            disabledContainerColor = palette.background.copy(alpha = 0.52f),
            disabledContentColor = palette.foreground.copy(alpha = 0.52f)
        )
    ) {
        LumenIcon(
            name = name,
            contentDescription = contentDescription,
            size = metrics.iconSize,
            tint = palette.foreground.copy(alpha = if (enabled) 1f else 0.52f)
        )
    }
}

@Immutable
data class LumenButtonMetrics(
    val horizontalPadding: Dp,
    val minHeight: Dp
) {
    companion object {
        fun resolve(size: LumenControlSize): LumenButtonMetrics = when (size) {
            LumenControlSize.Lg -> LumenButtonMetrics(LumenSpacing.Xl, 52.dp)
            LumenControlSize.Md -> LumenButtonMetrics(LumenSpacing.Lg, 44.dp)
            LumenControlSize.Sm -> LumenButtonMetrics(LumenSpacing.Md, 36.dp)
        }
    }
}

private data class LumenButtonPalette(
    val background: Color,
    val border: Color,
    val foreground: Color
)

private fun buttonPalette(
    colors: LumenColorPalette,
    intent: LumenButtonIntent
): LumenButtonPalette = when (intent) {
    LumenButtonIntent.Danger -> LumenButtonPalette(colors.danger, colors.danger, colors.onDanger)
    LumenButtonIntent.Primary -> LumenButtonPalette(colors.brandSolid, colors.brandSolid, colors.onBrand)
    LumenButtonIntent.Quiet -> LumenButtonPalette(Color.Transparent, Color.Transparent, colors.inkSoft)
    LumenButtonIntent.Secondary -> LumenButtonPalette(colors.surfaceMuted, colors.line, colors.ink)
}

@Composable
fun LumenButton(
    onClick: () -> Unit,
    modifier: Modifier = Modifier,
    intent: LumenButtonIntent = LumenButtonIntent.Primary,
    size: LumenControlSize = LumenControlSize.Md,
    loading: Boolean = false,
    enabled: Boolean = true,
    content: @Composable RowScope.() -> Unit
) {
    val palette = buttonPalette(LocalLumenTheme.current.colors, intent)
    val metrics = LumenButtonMetrics.resolve(size)
    val semanticsModifier = if (loading) {
        Modifier.semantics { stateDescription = "Loading" }
    } else {
        Modifier
    }

    Button(
        onClick = onClick,
        modifier = modifier
            .then(semanticsModifier)
            .defaultMinSize(minHeight = metrics.minHeight),
        enabled = enabled && !loading,
        shape = RoundedCornerShape(LumenRadius.Sm),
        border = BorderStroke(1.dp, palette.border),
        colors = ButtonDefaults.buttonColors(
            containerColor = palette.background,
            contentColor = palette.foreground,
            disabledContainerColor = palette.background.copy(alpha = 0.52f),
            disabledContentColor = palette.foreground.copy(alpha = 0.52f)
        ),
        contentPadding = PaddingValues(
            horizontal = metrics.horizontalPadding,
            vertical = LumenSpacing.Sm
        )
    ) {
        Row(
            horizontalArrangement = Arrangement.Center,
            verticalAlignment = Alignment.CenterVertically
        ) {
            if (loading) {
                CircularProgressIndicator(
                    modifier = Modifier.size(16.dp),
                    color = palette.foreground,
                    strokeWidth = 2.dp
                )
                Spacer(modifier = Modifier.width(LumenSpacing.Sm))
            }
            content()
        }
    }
}

@Composable
fun LumenTextField(
    value: String,
    onValueChange: (String) -> Unit,
    label: String,
    modifier: Modifier = Modifier,
    size: LumenControlSize = LumenControlSize.Md,
    error: Boolean = false,
    errorMessage: String? = null,
    enabled: Boolean = true,
    visualTransformation: VisualTransformation = VisualTransformation.None
) {
    val colors = LocalLumenTheme.current.colors
    val metrics = LumenButtonMetrics.resolve(size)

    OutlinedTextField(
        value = value,
        onValueChange = onValueChange,
        modifier = modifier.defaultMinSize(minHeight = metrics.minHeight),
        enabled = enabled,
        isError = error,
        singleLine = true,
        label = { Text(label) },
        supportingText = errorMessage?.let { message -> ({ Text(message) }) },
        visualTransformation = visualTransformation,
        shape = RoundedCornerShape(LumenRadius.Sm),
        colors = OutlinedTextFieldDefaults.colors(
            focusedContainerColor = colors.surface,
            unfocusedContainerColor = colors.surface,
            disabledContainerColor = colors.surface,
            focusedTextColor = colors.ink,
            unfocusedTextColor = colors.ink,
            focusedBorderColor = colors.brand,
            unfocusedBorderColor = colors.line,
            errorBorderColor = colors.danger,
            cursorColor = colors.brand
        )
    )
}

enum class LumenBadgeTone {
    Accent,
    Danger,
    Neutral,
    Success,
    Warning
}

@Composable
fun LumenBadge(
    text: String,
    modifier: Modifier = Modifier,
    tone: LumenBadgeTone = LumenBadgeTone.Neutral
) {
    val colors = LocalLumenTheme.current.colors
    val foreground = when (tone) {
        LumenBadgeTone.Accent -> colors.accent
        LumenBadgeTone.Danger -> colors.danger
        LumenBadgeTone.Neutral -> colors.inkSoft
        LumenBadgeTone.Success -> colors.success
        LumenBadgeTone.Warning -> colors.warning
    }
    val background = if (tone == LumenBadgeTone.Neutral) {
        colors.surfaceMuted
    } else {
        foreground.copy(alpha = 0.12f)
    }

    Surface(
        modifier = modifier,
        color = background,
        contentColor = foreground,
        shape = RoundedCornerShape(LumenRadius.Full)
    ) {
        Text(
            text = text,
            modifier = Modifier.padding(horizontal = LumenSpacing.Sm, vertical = LumenSpacing.Xs),
            style = MaterialTheme.typography.labelSmall.copy(fontWeight = FontWeight.SemiBold)
        )
    }
}

@Composable
fun LumenDivider(modifier: Modifier = Modifier) {
    val colors = LocalLumenTheme.current.colors

    HorizontalDivider(modifier = modifier, color = colors.line)
}

@Composable
fun LumenSpinner(
    modifier: Modifier = Modifier,
    label: String = "Loading"
) {
    val colors = LocalLumenTheme.current.colors

    CircularProgressIndicator(
        modifier = modifier.semantics { contentDescription = label },
        color = colors.brand
    )
}
